pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,220 @@
+
+	return r.handlePrivateEndpointInterface(ctx, projectID, akoPrivateEndpoint, akoPEService, atlasPEService)
+}
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointInterface(
+	ctx *workflow.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+	akoPEService privateendpoint.EndpointService,
+	atlasPEService privateendpoint.EndpointService,
+) (ctrl.Result, error) {
+	if len(akoPEService.EndpointInterfaces()) == 0 && len(atlasPEService.EndpointInterfaces()) == 0 {
+		return r.waitForConfiguration(ctx, akoPrivateEndpoint, atlasPEService)
+	}
+
+	inProgress := false
+	for _, akoPEInterface := range akoPEService.EndpointInterfaces() {
+		atlasPEInterfaces := atlasPEService.EndpointInterfaces()
+		atlasPEInterface := atlasPEInterfaces.Get(akoPEInterface.InterfaceID())
+		existInAtlas := atlasPEInterface != nil
+		inProgress = isInterfaceInProgress(akoPEInterface.Status())
+		var err error
+
+		if !existInAtlas {
+			gcpProjectID := getGCPProjectID(akoPrivateEndpoint, akoPEInterface.InterfaceID())
+			_, err = r.privateEndpointService.CreatePrivateEndpointInterface(
+				ctx.Context,
+				projectID,
+				akoPrivateEndpoint.Spec.Provider,
+				akoPrivateEndpoint.Status.ServiceID,
+				gcpProjectID,
+				akoPEInterface,
+			)
+			if err != nil {
+				return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToCreate, err)
+			}
+
+			inProgress = true
+		}
+	}
+
+	for _, atlasPEInterface := range atlasPEService.EndpointInterfaces() {
+		akoPEInterfaces := akoPEService.EndpointInterfaces()
+		akoPEInterface := akoPEInterfaces.Get(atlasPEInterface.InterfaceID())
+		wasDeleted := akoPEInterface == nil
+		inProgress = isInterfaceInProgress(atlasPEInterface.Status())
+		var err error
+
+		if wasDeleted && atlasPEInterface.Status() != privateendpoint.StatusDeleting {
+			err = r.privateEndpointService.DeleteEndpointInterface(ctx.Context, projectID, akoPrivateEndpoint.Spec.Provider, akoPrivateEndpoint.Status.ServiceID, atlasPEInterface.InterfaceID())
+			if err != nil {
+				return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToDelete, err)
+			}
+
+			inProgress = true
+
+			continue
+		}
+
+		if hasInterfaceFailed(atlasPEInterface.Status()) {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointFailedToConfigure, errors.New(atlasPEInterface.ErrorMessage()))
+		}
+	}
+
+	if !inProgress && reflect.DeepEqual(akoPEService.EndpointInterfaces(), atlasPEService.EndpointInterfaces()) {
+		return r.ready(ctx, akoPrivateEndpoint, atlasPEService)
+	}
+
+	return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointReady, workflow.PrivateEndpointUpdating, "Private Endpoints are being updated")
+}
+
+func (r *AtlasPrivateEndpointReconciler) getPrivateEndpointService(
```

Comment:
This is a bit too magic right?
It is searching by Service ID, if set, otherwise listing and matching by region. Is that second match alwasy accurate?

I would rename the method to be more explicit:

```suggestion
func (r *AtlasPrivateEndpointReconciler) getPrivateEndpointServiceByIDorRegion(
```

Reply:
the `Get` semantics for me is that it is supposed to return a single match or error the internal search is less relevant.
If I would be accepting an ID or region as parameters, then the renaming would make sense to me.

Reply:
Well I think this method does too much, I would rather have a search by id then a filter by region explicitly on the caller. So that is more obvious what is going on there. The code in the caller is still small (one if and else clause) and at this level we have the find by id and list and filter by region as 2 smaller functions.

Basically having methods that do one thing as said in the method name.

But maybe that is just me.